src/libPersistence/Retriever.cpp

```diff
@@ -236,11 +236,25 @@ bool Retriever::RetrieveBlockLink(bool trimIncompletedBlocks) {
 
   /// Check whether the termination of last running happens before the last
   /// DSEpoch properly ended.
-  bytes isDSIncompleted;
-  if (!BlockStorage::GetBlockStorage().GetMetadata(MetaType::DSINCOMPLETED,
-                                                   isDSIncompleted)) {
-    LOG_GENERAL(WARNING, "No GetMetadata or failed");
-    return false;
+  uint64_t epochFinNum = 0;
+  bool toDelete = false;
+  if (!BlockStorage::GetBlockStorage().GetEpochFin(epochFinNum)) {
+    LOG_GENERAL(WARNING, "BlockStorage::GetEpochFin failed");
+    // return false;
+    bytes isDSIncompleted;
+    if (!BlockStorage::GetBlockStorage().GetMetadata(MetaType::DSINCOMPLETED,
+                                                     isDSIncompleted)) {
+      LOG_GENERAL(WARNING, "No GetMetadata or failed");
+      return false;
+    }
+    if (isDSIncompleted[0] == '1') {
+      /// Removing incompleted DS for upgrading protocol
+      /// Keeping incompleted DS for node recovery
+      if (trimIncompletedBlocks) {
+        LOG_GENERAL(INFO, "Has incompleted DS Block, remove it");
+        toDelete = true;
+      }
+    }
   }
```

Comment:
do we need to handle `else` here, return false?

Reply:
If `else` then we will just ignore, no special action needed